### DISCUSSION TITLE
How to convert tap match() to node:test match()?

### DISCUSSION THREAD

**Discussion by sourcecodeit** [fastify/fastify#5628] [General] [answered]
Hi,

I'm starting to have a look at migrating plugins to `node:test` but I'm not sure how to convert tap `match()` function as the `node:test` `match()` is completely different and I don't see a similar function in `node:test`.

Example:

current code:

```
t.assert.match(res.headers, {
  'access-control-allow-origin': '*'
})
```

replaced with:

```
t.assert.match(JSON.stringify(res.headers), /"access-control-allow-origin":"*"/)
```

which doesn't look very nice. Any idea?

**Comment by Uzlopak** [comment]
`match` from tap is in this case a `t.assert.deepEqual`

better do just

t.assert.strictEqual(res.headers?.['access-control-allow-origin'], '*')

not tested.

**Reply by sourcecodeit** [reply]
> t.assert.strictEqual(res.headers?.['access-control-allow-origin'], '*')
> 
> not tested.

It works. So for cases like this:

```
    t.match(res.headers, {
      'access-control-allow-origin': '*',
      'content-length': statSync(filePath).size
    })
```
it would be split in 2:

```
    t.assert.strictEqual(res.headers?.['access-control-allow-origin'], '*')
    t.assert.equal(res.headers?.['content-length'], statSync(filePath).size //not strict here
```

it works. What do you think, any better way to handle that?

**Reply by Uzlopak** [reply]
You could try deepEqual

**Reply by sourcecodeit** [reply]
> You could try deepEqual

Not working:
`
      t.assert.deepEqual(res.headers, { 'access-control-allow-origin' : '*' })
`
``` 
      Expected values to be loosely deep-equal:

      {
        'access-control-allow-origin': '*',
        'content-length': '21483',
        'content-type': 'application/json',
        connection: 'keep-alive'
      }

      should loosely deep-equal

      {
        'access-control-allow-origin': '*'
      }
    code: ERR_ASSERTION
```

**Reply by Uzlopak** [reply]
ok, i am seeing now that deepEqual is an alias for deepStrictEqual. So yeah than you have to do it as you did before.

**Reply by jsumners** [reply]
There is not an equivalent. It'd be cool if someone ported it to core, though.

**Reply by Eomm** [reply]
We should follow: https://github.com/nodejs/node/pull/53415

**Reply by jsumners** [reply]
That is also not an equivalent for https://github.com/tapjs/tapjs/blob/8f2baa7eb5e1f4cd0fcf8b60da4fb4e41ea680f1/docs/src/content/docs/api/asserts/index.md?plain=1#L318-L338

**Comment by jsumners** [comment]
There is not an equivalent and it's rather disappointing.

**Comment by mcollina** [comment]
@mateonunez do you think you could wrap https://github.com/platformatic/platformatic/blob/main/packages/utils/lib/match.js into an assertion that we could use as a module? I would also open the same PR to Node.js core.

**Reply by mateonunez** [reply]
Hey @mcollina, yes, it could be done it that way so we can have the same behaviors as the tap assertion.

Do you suggest to integrate the match function directly into the Node.js assertion library (and then refactor on plt) or somewhere?

**Reply by mcollina** [reply]
I think we need a library _right now_, and a PR to Node.js too (but given LTS, this will have a slower adoption).

**Reply by jsumners** [reply]
Particularly with the lack of backports to 18.

**Reply by mateonunez** [reply]
I did this in a few minutes https://github.com/mateonunez/matchare, in the next few hours I will open a PR on Node.js to include this lib as an assertion. 

This lib may change until the stable version is released.

**Reply by jsumners** [reply]
I wouldn't simply state "follows the same match pattern as node-tap." People may be unfamiliar, there isn't a link, and if there were a link it will likely expire. It would be far better to fully document the module in the readme.

Also, that module seems to be ESM only.

**Reply by mcollina** [reply]
@mateonunez I don't
think an esm-only module would work for us.

**Reply by mateonunez** [reply, marked as answer]
The lib now ships on cjs and esm (v0.0.2), I'm improving the documentation (this will produce the stable version). Unfortunately, I can't allocate the time needed to bring this in quickly on Node.js. But I'm working on it, anyone who would like to contribute in this is welcome! @mcollina

**Reply by mateonunez** [reply]
If no impediments I’m going to submit the Node PR today. A couple of tests and it will be done
